File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public static boolean canShip(int weights[], int shipWeight, int daysHave) {
        int daysNeeded = 1;
        int tSumSoFar = 0;
        for (int weight: weights) {
            if (tSumSoFar + weight <= shipWeight) {
                tSumSoFar += weight;
            }
            else {
                daysNeeded++;
                tSumSoFar = weight;
            }
        }
        return daysNeeded <= daysHave;
    }
    public static int getMax(int[] weights) {
        int maxWeight = weights[0];
        for (int weight: weights) {
            maxWeight = Math.max(weight, maxWeight);
        }
        return maxWeight;
    }
    public static int getSum(int[] weights) {
        int sum = 0;
        for (int weight: weights) {
            sum += weight;
        }
        return sum;
    }
    public int shipWithinDays(int[] weights, int days) {
        int low = getMax(weights), high = getSum(weights);
        while (low < high) {
            int mid = (low + high) / 2;
            if (canShip(weights, mid, days)) {
                high = mid;
            }
            else {
                low = mid + 1;
            }
        }
        return low;
    }
}

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
def canShip(weights, days_have, capacity):
    # Find the days_needed to ship all the weights under choosen capacity
    days_needed = 1
    cWeightSum = 0
    for w in weights:
        if cWeightSum + w <= capacity:
            cWeightSum += w
        else:
            days_needed += 1
            cWeightSum = w
    # Compare days_needed <= days_have (Capacity is a valid choice)
    return days_needed <= days_have

class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low = max(weights)
        high = sum(weights)
        while low < high:
            mid = (low + high) // 2
            if canShip(weights, days, mid):
                high = mid
            else:
                low = mid + 1
        return low
